File: packages/swt3-ai-dotnet/Fingerprint.cs
using System.Security.Cryptography;
using System.Text;

namespace Swt3Ai;

/// <summary>
/// SWT3 fingerprint minting and SHA-256 hashing utilities.
/// All methods produce output identical to the Python and TypeScript SDKs.
/// </summary>
public static class Fingerprint
{
    /// <summary>
    /// Mint an SWT3 fingerprint from the canonical formula.
    /// Returns the first 12 hex characters of SHA-256 applied to the canonical input string.
    /// </summary>
    public static string MintFingerprint(
        string tenantId,
        string procedureId,
        double factorA,
        double factorB,
        double factorC,
        long timestampMs)
    {
        var input = $"WITNESS:{tenantId}:{procedureId}:{NumStr(factorA)}:{NumStr(factorB)}:{NumStr(factorC)}:{timestampMs}";
        return Sha256Hex(input, 12);
    }

    /// <summary>
    /// Compute a truncated SHA-256 hash. Default length is 16 hex characters.
    /// Used for prompt/response hashing (16), fingerprints (12), and full digests (64).
    /// </summary>
    public static string Sha256Truncated(string data, int length = 16)
    {
        return Sha256Hex(data, length);
    }

    /// <summary>
    /// Compute SHA-256 and return the first N hex characters.
    /// </summary>
    public static string Sha256Hex(string data, int length = 64)
    {
        var bytes = Encoding.UTF8.GetBytes(data);
        var hash = SHA256.HashData(bytes);
        var hex = Convert.ToHexString(hash).ToLowerInvariant();
        return hex[..Math.Min(length, 64)];
    }

    /// <summary>
    /// Format a numeric factor as a string matching the canonical formula.
    /// Integer-valued doubles are formatted without decimals: 1.0 -> "1"
    /// </summary>
    private static string NumStr(double v)
    {
        if (double.IsFinite(v) && v == Math.Floor(v))
        {
            return ((long)v).ToString();
        }
        return v.ToString();
    }
}

File: packages/swt3-ai-ruby/lib/swt3_ai/fingerprint.rb
require "openssl"

module Swt3Ai
  module Fingerprint
    # Mint an SWT3 fingerprint from the canonical formula.
    #
    # Returns the first 12 hex characters of:
    #   SHA-256("WITNESS:{tenant}:{proc}:{fa}:{fb}:{fc}:{ts_ms}")
    #
    # This formula is locked and must produce identical output across
    # all SWT3 SDK implementations (Python, TypeScript, Rust, C#, Ruby).
    def self.mint_fingerprint(tenant_id, procedure_id, factor_a, factor_b, factor_c, timestamp_ms)
      input = "WITNESS:#{tenant_id}:#{procedure_id}:#{num_str(factor_a)}:#{num_str(factor_b)}:#{num_str(factor_c)}:#{timestamp_ms}"
      sha256_hex(input, 12)
    end

    # Compute a truncated SHA-256 hash. Default length is 16 hex characters.
    def self.sha256_truncated(data, length = 16)
      sha256_hex(data, length)
    end

    # Compute SHA-256 and return the first N hex characters.
    def self.sha256_hex(data, length = 64)
      digest = OpenSSL::Digest::SHA256.hexdigest(data.to_s)
      digest[0, [length, 64].min]
    end

    # Get the current timestamp in milliseconds and epoch seconds.
    def self.timestamp_ms
      ms = (Time.now.to_f * 1000).to_i
      epoch = ms / 1000
      [ms, epoch]
    end

    # Format a numeric factor as a string matching the canonical formula.
    # Integer-valued floats are formatted without decimals: 1.0 -> "1"
    def self.num_str(v)
      v == v.to_i ? v.to_i.to_s : v.to_s
    end

    private_class_method :num_str
  end
end
